### ISSUE TITLE
Tune the code for specific data

### ISSUE THREAD

**Issue by gnostis** [emedvedev/attention-ocr#133]
My input is formatted codes where first letter is always lowercase and 5 digits after:
a00000
Is it possible to limit recognition for overall better efficiency?
By default I can only use --max-prediction to limit length. But my goal is to set it to exact length and format. Could you point me right way how to better achieve this?

**Comment by emedvedev** [comment]
Not without significant changes to the model and the code, I’m afraid, and the accuracy gain may not be worth it. In my experience, the model works quite well with images of the same format (e.g. captcha) when properly trained as it is.

Do you actually see poor performance, or is it mostly a hypothetical question? Some hyperparameter tuning or a better dataset could be enough to improve accuracy if you’re not getting satisfying results.

**Comment by gnostis** [comment]
The results are good, only about 2% misses. All samples are quite noisy. Although these errors occur on quite simple copies.  I tried to double the data set to 40,000, but it gave a much worse result. Perhaps there is bad input data.  Therefore, I wanted to consider all possible options to influence the result because it's critical. I'll try to play with the hyperparameters and maybe preprocessing as last resort.

**Comment by gnostis** [comment]
During testing of the model after each checkpoint, I noticed that the model was learning characters in alphabetical order.  I mean, first, the model "knows" only the number 1 and recognizes all the examples as 111111, then the letter A is added, and we have something like A11A11, then B is added, and so on.  Is this how it should be?  I thought that if we randomly chose a batch, then all letters would get there and the recognition would also correlate with the whole list of characters, but with a minimum probability. 

My data set was initially 20000, then 40000, now 120000, but the best results were achieved by 20000 and worsened with an increase in the data set. Tuning batch size has almost no effect.

**Comment by emedvedev** [comment]
About the larger dataset: it could be that you're overtraining: too many epochs, wrong hyperparameters, the testing dataset is significantly different from your training dataset and the model is unable to generalize, etc. Sadly, I can't help there, as it requires working with the specific dataset, and the model is unlikely to be at fault here.

About the order: during the first few steps, it simply tries the "guesses" in the order of the [character map](https://github.com/emedvedev/attention-ocr/blob/master/aocr/util/data_gen.py#L23), which is `0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ`, so it's not surprising and not wrong either. :) It's a good observation, but pretty much expected behavior.

**Comment by AhmadShaik** [comment]
> My input is formatted codes where first letter is always lowercase and 5 digits after:
> a00000
> Is it possible to limit recognition for overall better efficiency?
> By default I can only use --max-prediction to limit length. But my goal is to set it to exact length and format. Could you point me right way how to better achieve this?

@gnostis I am also working on the similar problem. I want to recognise a label number that contains one uppercase character and 6 to 8 digits. It would be really helpful if you share details like how you prepared your dataset, how many epochs you trained the model, what type of fonts you used.
